### ISSUE TITLE
[BUG] run unit test shows 'Cannot re-initialize CUDA in forked subprocess'

### ISSUE THREAD

**Issue by delock** [microsoft/DeepSpeed#1987]
**Describe the bug**
When run unit test test_zero2_reduce_scatter_off, get error message: "Cannot re-initialize CUDA in forked subprocess"

**To Reproduce**
1. Go to tests/unit
2. run command 'pytest -k test_zero2_reduce_scatter_off'
3. The test would fail and see message "RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method"
4. Most of unit test have same error message when they have @distributed_test decorator
5. Manually modify common.py to use spawn method does not solve the problem, because in spawn cannot pickle local function

**Expected behavior**
The test would not fail with error message "RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method"

**ds_report output**
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
 [WARNING]  please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [NO] ....... [NO]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/guizili/miniconda3/envs/ds/lib/python3.9/site-packages/torch']
torch version .................... 1.11.0+cu115
torch cuda version ............... 11.5
torch hip version ................ None
nvcc version ..................... 11.6
deepspeed install path ........... ['/home/guizili/frameworks.ai.benchmarking.other.deepspeed/deepspeed']
deepspeed info ................... 0.6.6+3da84185, 3da84185, upmaster
deepspeed wheel compiled w. ...... torch 1.11, cuda 11.5
```

**Screenshots**

**System info (please complete the following information):**
 - OS: Ubuntu 20.04
 - GPU count and types: one machine with x1 A100
 - Interconnects (if applicable): N/A
 - Python version: 3.9.7
 - Any other relevant info about your setup

**Launcher context**
N/A, I'm running unit test

**Docker context**
N/A

**Additional context**

**Comment by delock** [comment]
also observed the same when using 'pip install deepspeed', which installs deepspeed 0.6.5 and get same error.

**Comment by delock** [comment]
Suspect this failure is caused by tests/unit/util.py:bf16_requried_version_check(), the torch runtime call would initialize CUDA before fork, which cause unit test to fail.
```
def bf16_required_version_check():
    TORCH_MAJOR = int(torch.__version__.split('.')[0])
    TORCH_MINOR = int(torch.__version__.split('.')[1])

    if type(torch.cuda.nccl.version()) != tuple:
        return False
    else:
        NCCL_MAJOR = torch.cuda.nccl.version()[0]
        NCCL_MINOR = torch.cuda.nccl.version()[1]

    CUDA_MAJOR = int(torch_info['cuda_version'].split('.')[0])
    if (TORCH_MAJOR > 1 or
        (TORCH_MAJOR == 1 and TORCH_MINOR >= 10)) and (CUDA_MAJOR >= 11) and (
            NCCL_MAJOR > 2 or
            (NCCL_MAJOR == 2 and NCCL_MINOR >= 10)) and **torch.cuda.is_bf16_supported():**
        return True
    else:
        return False
```

**Comment by mrwyattii** [comment]
@delock Thanks for bringing this to our attention and doing some work to debug it. I think you are correct and I'm working on a fix. We will also add a CI workflow with the latest torch version to catch this in the future!

**Comment by mrwyattii** [comment]
This bug will be resolved when #1990 merges
